Exceptions thrown from classes executed by the java api are not propagated

When a java api operation throws an exception, the exception message and stacktrace are not shown as part of the scenario stacktrace, making it difficult to debug.

**To reproduce:**
- Start from a new branch
- Edit `org.jsmart.zerocode.core.AddService` and replace `return 30` in the 'anInteger' method by `throw new RuntimeException("INTENTIONAL EXCEPTION")`
- Optionally, edit the resource file `unit_test_files/java_apis/02_test_json_java_service_method_no_param.json` and remove `"loop": 5,`
- Run the test in package `org.jsmart.zerocode.core.javamethod`

**Expected:** Some information about the RuntimeException `INTENTIONAL EXCEPTION`, and its stacktrace should be shown
**Actual:** It is ommited. This is the stacktrace:

```
2024-10-20 11:55:44,184 [main] ERROR org.jsmart.zerocode.core.engine.executor.javaapi.JavaMethodExecutorImpl - Java exec(): Invocation failed for method anInteger in class org.jsmart.zerocode.core.AddService. Exception - java.lang.reflect.InvocationTargetException
2024-10-20 11:55:44,184 [main] ERROR org.jsmart.zerocode.core.engine.executor.javaapi.JavaMethodExecutorImpl - Exception - java.lang.RuntimeException: Java exec(): Invocation failed for method anInteger in class org.jsmart.zerocode.core.AddService
2024-10-20 11:55:44,185 [main] ERROR org.jsmart.zerocode.core.engine.executor.ApiServiceExecutorImpl - Java method execution exception - java.lang.RuntimeException: java.lang.RuntimeException: Java exec(): Invocation failed for method anInteger in class org.jsmart.zerocode.core.AddService
java.lang.RuntimeException: java.lang.RuntimeException: java.lang.RuntimeException: Java exec(): Invocation failed for method anInteger in class org.jsmart.zerocode.core.AddService
	at org.jsmart.zerocode.core.engine.executor.ApiServiceExecutorImpl.executeJavaOperation(ApiServiceExecutorImpl.java:53)
	at org.jsmart.zerocode.core.runner.ZeroCodeMultiStepsScenarioRunnerImpl.executeApi(ZeroCodeMultiStepsScenarioRunnerImpl.java:471)
	at org.jsmart.zerocode.core.runner.ZeroCodeMultiStepsScenarioRunnerImpl.executeRetry(ZeroCodeMultiStepsScenarioRunnerImpl.java:256)
	at org.jsmart.zerocode.core.runner.ZeroCodeMultiStepsScenarioRunnerImpl.executeRetryWithSteps(ZeroCodeMultiStepsScenarioRunnerImpl.java:195)
	at org.jsmart.zerocode.core.runner.ZeroCodeMultiStepsScenarioRunnerImpl.executeSteps(ZeroCodeMultiStepsScenarioRunnerImpl.java:177)
	at org.jsmart.zerocode.core.runner.ZeroCodeMultiStepsScenarioRunnerImpl.runScenario(ZeroCodeMultiStepsScenarioRunnerImpl.java:140)
(...)
Caused by: java.lang.RuntimeException: java.lang.RuntimeException: Java exec(): Invocation failed for method anInteger in class org.jsmart.zerocode.core.AddService
	at org.jsmart.zerocode.core.engine.executor.javaapi.JavaMethodExecutorImpl.execute(JavaMethodExecutorImpl.java:63)
	at org.jsmart.zerocode.core.engine.executor.ApiServiceExecutorImpl.executeJavaOperation(ApiServiceExecutorImpl.java:50)
	... 22 more
Caused by: java.lang.RuntimeException: Java exec(): Invocation failed for method anInteger in class org.jsmart.zerocode.core.AddService
	at org.jsmart.zerocode.core.engine.executor.javaapi.JavaMethodExecutorImpl.executeWithParams(JavaMethodExecutorImpl.java:114)
	at org.jsmart.zerocode.core.engine.executor.javaapi.JavaMethodExecutorImpl.execute(JavaMethodExecutorImpl.java:47)
	... 23 more
2024-10-20 11:55:44,186 [main] ERROR org.jsmart.zerocode.core.runner.ZeroCodeMultiStepsScenarioRunnerImpl - ###Exception while executing a step in the zerocode dsl.
2024-10-20 11:55:44,188 [main] ERROR org.jsmart.zerocode.core.runner.StepNotificationHandler - Exception occurred while executing Scenario:[Given_When_Then-Flow name For Java Service], --> Step:[javaMethodExecution], Details: java.lang.RuntimeException: ZeroCode Step execution failed. Details:java.lang.RuntimeException: java.lang.RuntimeException: java.lang.RuntimeException: Java exec(): Invocation failed for method anInteger in class org.jsmart.zerocode.core.AddService
2024-10-20 11:55:44,193[main] org.jsmart.zerocode.core.runner.ZeroCodeMultiStepsScenarioRunnerImpl -m
```

I've located the cause: `org.jsmart.zerocode.core.engine.executor.javaapi.executeWithParams` catches exceptions and rethrows a new RuntimeException with the message only.

If adding both the message and catched exception to the throw statement, the stacktrace is augmented with: 
```
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jsmart.zerocode.core.engine.executor.javaapi.JavaMethodExecutorImpl.executeWithParams(JavaMethodExecutorImpl.java:107)
	... 24 more
Caused by: java.lang.RuntimeException: INTENTIONAL EXCEPTION
	at org.jsmart.zerocode.core.AddService.anInteger(AddService.java:27)
	... 29 more
```

I will prepare a PR to fix it. @authorjapps can you please assign this to me?

> I've located the cause: `org.jsmart.zerocode.core.engine.executor.javaapi.executeWithParams` catches exceptions and rethrows a new RuntimeException with the message only.
> 
> If adding both the message and catched exception to the throw statement, the stacktrace is augmented with:
> 
> ```
> Caused by: java.lang.reflect.InvocationTargetException
> 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at java.lang.reflect.Method.invoke(Method.java:498)
> 	at org.jsmart.zerocode.core.engine.executor.javaapi.JavaMethodExecutorImpl.executeWithParams(JavaMethodExecutorImpl.java:107)
> 	... 24 more
> Caused by: java.lang.RuntimeException: INTENTIONAL EXCEPTION
> 	at org.jsmart.zerocode.core.AddService.anInteger(AddService.java:27)
> 	... 29 more
> ```
> 
> I will prepare a PR to fix it. @authorjapps can you please assign this to me?

yep, that's right. Sounds amazing. Please go ahead 🚀 
Issue assigned to you now.

Also, added you now as collaborator. Click [here](https://github.com/authorjapps/zerocode/invitations) to accept.
Now, you can, 
- Approve PRs
- Trigger CI/CD pipelines
- assign/unassign issues
etc

Also, can you add relevant logs as WARN level for the DB executor, to surface important exception messages etc to the end user? This will be very helpful to Devs and SDETs to avoid rabbit holes when any problem arises.

_(Ignore me if you've already thought of doing it anyways, as part of this ticket :) )_

--- a/core/src/main/java/org/jsmart/zerocode/core/AddService.java
+++ b/core/src/main/java/org/jsmart/zerocode/core/AddService.java
@@ -21,6 +21,12 @@ public class AddService {
         return myNumber.getNumber() * myNumber.getNumber();
     }
 
+    public Double squareRoot(Double number) {
+    	if (number < 0.0)
+    		throw new RuntimeException("Can not square root a negative number");
+        return Math.sqrt(number);
+    }
+    
     public Integer anInteger() {
         logger.debug("Returning a number ");
 

--- a/core/src/main/java/org/jsmart/zerocode/core/engine/executor/javaapi/JavaMethodExecutorImpl.java
+++ b/core/src/main/java/org/jsmart/zerocode/core/engine/executor/javaapi/JavaMethodExecutorImpl.java
@@ -111,7 +111,7 @@ public class JavaMethodExecutorImpl implements JavaMethodExecutor {
         } catch (Exception e) {
             String errMsg = format("Java exec(): Invocation failed for method %s in class %s", methodName, qualifiedClassName);
             LOGGER.error(errMsg + ". Exception - " + e);
-            throw new RuntimeException(errMsg);
+            throw new RuntimeException(errMsg, e);
         }
     }
 

--- a/core/src/test/java/org/jsmart/zerocode/core/engine/executor/javaapi/JavaMethodExecutorImplTest.java
+++ b/core/src/test/java/org/jsmart/zerocode/core/engine/executor/javaapi/JavaMethodExecutorImplTest.java
@@ -3,6 +3,7 @@ package org.jsmart.zerocode.core.engine.executor.javaapi;
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.google.inject.Guice;
 import com.google.inject.Injector;
+import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
 import java.util.List;
 import org.jsmart.zerocode.core.di.main.ApplicationMainModule;
@@ -16,6 +17,9 @@ import org.junit.rules.ExpectedException;
 
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.core.Is.is;
+import static org.hamcrest.core.StringContains.containsString;
+import static org.hamcrest.CoreMatchers.notNullValue;
+import static org.junit.Assert.assertThrows;
 
 public class JavaMethodExecutorImplTest {
 
@@ -122,6 +126,29 @@ public class JavaMethodExecutorImplTest {
         Object result = methodExecutor.executeWithParams(serviceName, methodName, request);
 
         assertThat(result, is(900));
+    }
+
+    @Test
+    public void willPropagateExceptions() throws Exception {
+        String scenariosJsonAsString = SmartUtils.readJsonAsString("unit_test_files/java_apis/03_test_json_java_service_method_Exception.json");
+        final ScenarioSpec scenarioSpec = smartUtils.getMapper().readValue(scenariosJsonAsString, ScenarioSpec.class);
+
+        String serviceName = scenarioSpec.getSteps().get(0).getUrl();
+        String methodName = scenarioSpec.getSteps().get(0).getOperation();
+        String requestJson = scenarioSpec.getSteps().get(0).getRequest().toString();
+        List<Class<?>> argumentTypes = methodExecutor.getParameterTypes(serviceName, methodName);
+
+        Object request = mapper.readValue(requestJson, argumentTypes.get(0));
+        
+        RuntimeException exception = assertThrows(RuntimeException.class, () -> {
+        	methodExecutor.executeWithParams(serviceName, methodName, request);
+		});
+        exception.printStackTrace();
+        assertThat(exception.getMessage(), containsString("Invocation failed for method squareRoot"));
+        // The target exception is included in this exception (inside of the InvocationTargetException)
+        assertThat(exception.getCause(), is(notNullValue()));
+        assertThat(((InvocationTargetException)exception.getCause()).getTargetException().getMessage(), 
+        		is("Can not square root a negative number"));
     }
 
     @Test

--- a/core/src/test/resources/unit_test_files/java_apis/03_test_json_java_service_method_Exception.json
+++ b/core/src/test/resources/unit_test_files/java_apis/03_test_json_java_service_method_Exception.json
@@ -1,0 +1,13 @@
+{
+  "scenarioName": "Given_When_Then-Flow name For Java Service",
+  "steps": [
+    {
+      "name": "javaMethodException",
+      "url": "org.jsmart.zerocode.core.AddService",
+      "operation": "squareRoot",
+      "request": "-255.0", //<-- This negative number should throw an exception
+      "assertions": {
+      }
+    }
+  ]
+}

